File: Chapter5Hook11/app/src/main/java/com/mars/chapter5hook11/MarsInstrumentation.java
package com.mars.chapter5hook11;

import android.app.Activity;
import android.app.Instrumentation;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;

import com.mars.chapter5hook11.util.MethodUtils;

/**
 * Created by geyan on 2020/12/17
 */
public class MarsInstrumentation extends Instrumentation {


    // ActivityThread中原始的对象, 保存起来
    Instrumentation mBase;

    public MarsInstrumentation(Instrumentation base) {
        mBase = base;
    }

    public ActivityResult execStartActivity(
            Context who, IBinder contextThread, IBinder token, Activity target,
            Intent intent, int requestCode, Bundle options) {
        Log.e("gy", "MarsInstrumentation---execStartActivity");
        // 反射调用原始的execStartActivity
        Class[] p1 = {Context.class, IBinder.class,
                IBinder.class, Activity.class,
                Intent.class, int.class, Bundle.class};
        Object[] v1 = {who, contextThread, token, target,
                intent, requestCode, options};
        return (ActivityResult) MethodUtils.invokeMethod(
                mBase, "execStartActivity", p1, v1);
    }
}

File: chapter5/Hook11/src/main/java/com/mars/hello/first/plugins/hook11/MarsInstrumentation.kt
package com.mars.hello.first.plugins.hook11

import android.app.Activity
import android.app.Instrumentation
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.IBinder
import com.mars.components.util.reflect.MethodUtils
import com.mars.hello.first.plugins.common.base.Logger

/**
 * Created by geyan on 2021/6/19
 *
 * hook Instrumentation
 */
class MarsInstrumentation(private val originInstrumentation: Instrumentation?) : Instrumentation() {

    companion object {
        const val TAG = "MarsInstrumentation"
    }

    fun execStartActivity(
        who: Context?, contextThread: IBinder?, token: IBinder?, target: Activity?,
        intent: Intent?, requestCode: Int, options: Bundle?
    ): ActivityResult? {
        Logger.e(TAG, "invoke execStartActivity")

        // 反射调用原始的execStartActivity
        val paramTypes = arrayOf(
            Context::class.java, IBinder::class.java,
            IBinder::class.java, Activity::class.java,
            Intent::class.java, Int::class.javaPrimitiveType, Bundle::class.java
        )

        val paramValues = arrayOf(
            who, contextThread, token, target,
            intent, requestCode, options
        )
        return MethodUtils.invokeMethod(
            originInstrumentation,
            "execStartActivity",
            paramTypes,
            paramValues
        ) as ActivityResult?
    }
}